druid-shell/src/alert.rs

```diff
@@ -0,0 +1,451 @@
+///
+/// While an alert dialog is open it takes interaction priority in its modality scope
+/// and the user won't be able to continue interacting with anything else.
+/// In window scope this means the user can't interact with the parent window
+/// and in application scope this means the user can't interact with any other window.
+///
+/// # Multiple alerts
+///
+/// There is no guaranteed behavior when multiple alert dialogs are requested with
+/// overlapping modality scope. Depending on platform specifics, all the alerts
+/// might be visible or just some of them. Regardless of visibility, only one of
+/// the alert dialogs will be interactable. When that alert dialog is closed,
+/// another one will become interactable.
+///
+/// If the modality scopes don't overlap, i.e. when multiple alerts are originating
+/// from different windows and they are all window scoped, all the alerts can be
+/// interacted with and they won't block eachother.
+#[derive(Debug, Clone)]
+pub struct AlertOptions {
+    /// Whether the alert is app-modal.
+    pub(crate) app_modal: bool,
+    /// The context of the alert.
+    pub(crate) context: String,
+    /// The primary message of the alert.
+    pub(crate) message: String,
+    /// The supplemental text of the alert.
+    pub(crate) description: String,
+    /// The icon of the alert.
+    pub(crate) icon: Option<AlertIcon>,
+    /// The primary button of the alert.
+    pub(crate) primary: AlertButton,
+    /// The cancel button of the alert.
+    pub(crate) cancel: Option<AlertButton>,
+    /// The alternative buttons of the alert.
+    pub(crate) alternatives: Vec<AlertButton>,
+}
+
+/// Alert dialog icon.
+///
+/// # macOS
+///
+/// On macOS the information and warning icons are the same.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum AlertIcon {
+    /// The alert dialog is providing important time sensitive information.
+    Information,
+    /// The alert dialog is warning about a potential future problem.
+    Warning,
+    /// The alert dialog is informing of a critical problem that has already happened.
+    Error,
+}
+
+/// A specific button for the alert dialog.
+///
+/// If the alert is just reporting information and it only needs a single button,  
+/// then you should use a primary button with the label **OK** or a translation of it.
+///
+/// The cancel button should always be labeled **Cancel** or a translation of it.
+///
+/// In all other cases you should give a short but descriptive label that includes a verb.
+/// The label should let the user know what will happen when they click the button.
+/// Assume that the user didn't actually read the alert message. This assumption ends up
+/// true more often than not, because alerts by design appear unexpectedly while the user
+/// is trying to accomplish some other task.
+///
+/// Labels like **Send message**, **View all**, **Reconnect**, **Remind me tomorrow**, and
```

Comment:
I might also include links to things like the macOS human interface guidelines.

Reply:
These links tend to 404 annoyingly fast and we also don't want to leave the impression that following any one platform's guidelines will lead to good results everywhere.

That said, I'm not trying to *hide* this information so I'll add a section linking to Apple/Microsoft/Gnome guidelines with a note about cross-platform considerations.